Testnet/StratisSwap/StratisSwap/JsonConfig/JsonConfig.cs

```diff
@@ -0,0 +1,157 @@
+    public bool IsContributor(Address address)
+    {
+        return PersistentState.GetBool($"{ContributorKey}:{address}");
+    }
+
+    /// <summary>
+    /// Updates the status of an admin. Requires admin privileges.
+    /// </summary>
+    /// <param name="address">The address to update.</param>
+    /// <param name="value">Status as boolean, true for approved admin, false for not admin.</param>
+    public void UpdateAdmin(Address address, bool value)
+    {
+        Assert(IsAdmin(Message.Sender));
+
+        UpdateAdminExecute(address, value);
+    }
+
+    private void UpdateAdminExecute(Address address, bool value)
+    {
+        PersistentState.SetBool($"{AdminKey}:{address}", value);
+
+        Log(new RoleLog
+        {
+            Blame = Message.Sender,
+            UpdatedAddress = address,
+            UpdatedValue = value,
+            Action = nameof(UpdateAdmin),
+            Block = Block.Number
+        });
+    }
+
+    /// <summary>
+    /// Updates the status of a contributor. Requires admin privileges.
+    /// </summary>
+    /// <param name="address">The address to update.</param>
+    /// <param name="value">Status as boolean, true for approved contributor, false for not contributor.</param>
+    public void UpdateContributor(Address address, bool value)
+    {
+        Assert(IsAdmin(Message.Sender));
+
+        PersistentState.SetBool($"{ContributorKey}:{address}", value);
+
+        Log(new RoleLog
+        {
+            Blame = Message.Sender,
+            UpdatedAddress = address,
+            UpdatedValue = value,
+            Action = nameof(UpdateContributor),
+            Block = Block.Number
+        });
+    }
+
+    /// <summary>
+    /// Adds a new configuration receipt to the contract.
+    /// </summary>
+    /// <param name="config">JSON payload as string.</param>
+    public void UpdateConfig(string config)
+    {
+        Assert(IsAdmin(Message.Sender) || IsContributor(Message.Sender));
+
+        UpdateConfigExecute(config);
+    }
+
+    private void UpdateConfigExecute(string config)
+    {
+        Log(new ConfigLog
+        {
+            Config = config,
+            Blame = Message.Sender,
+            Block = Block.Number
+        });
+    }
+
+    public struct RoleLog
+    {
+        /// <summary>
+        /// The address responsible for the role update.
+        /// </summary>
+        [Index]
+        public Address Blame;
+
+        /// <summary>
+        /// The address who's role was updated.
+        /// </summary>
+        [Index]
+        public Address UpdatedAddress;
+
+        /// <summary>
+        /// The action taken (e.g. UpdateAdmin or UpdateContributor)
+        /// </summary>
+        public string Action;
+
+        /// <summary>
+        /// The boolean value the status was updated too.
+        /// </summary>
+        public bool UpdatedValue;
+
+        /// <summary>
+        /// The block the update occurred.
+        /// </summary>
+        public ulong Block;
```

Comment:
As above, block number is already stored with receipt also.

Reply:
Good call but logging here saves the front end and API call because the raw block number doesn’t come back with the receipt search.